go/mysql/query.go

```diff
@@ -303,10 +308,35 @@ func (c *Conn) parseRow(data []byte, fields []*querypb.Field, reader func([]byte
 //	 2. if the server closes the connection when a command is in flight,
 //	    readComQueryResponse will fail, and we'll return CRServerLost(2013).
 func (c *Conn) ExecuteFetch(query string, maxrows int, wantfields bool) (result *sqltypes.Result, err error) {
-	result, _, err = c.ExecuteFetchMulti(query, maxrows, wantfields)
+	result, more, err := c.ExecuteFetchMulti(query, maxrows, wantfields)
+	if more {
+		// Multiple results are unexpected. Prioritize this "unexpected" error over whatever error we got from the first result.
+		err = errors.Join(ErrExecuteFetchMultipleResults, err)
+	}
+	// draining to make the connection clean.
+	err = c.drainMoreResults(more, err)
 	return result, err
 }
 
+// ExecuteFetchMultiDrain is for executing multiple statements in one call, but without
+// caring for any results. The function returns an error if any of the statements fail.
+// The function drains the query results of all statements, even if there's an error.
+func (c *Conn) ExecuteFetchMultiDrain(query string) (err error) {
+	_, more, err := c.ExecuteFetchMulti(query, 0, false)
```

Comment:
@shlomi-noach is the 0 here correct or could that result in a similar issue?

Reply:
It's correct, because the first query runs through `ReadQueryResult` which drains the results. Then, `ExecuteFetchMultiDrain` actively drains the results of the _remaining_ queries. What;'s important here is that we don't do double-draining on the same result, which is what happened previously and was fixed both my my change and then by @harshit-gangal 's change on top.

Reply:
We can set `-1` for visual consistency with other calls, too. That would be fine.

Reply:
I'm trying to make this work but keep getting the double-drain problem. @harshit-gangal is there a way for us to use `0` so that we don't read excessive rows into memory, and still drain correctly?

Reply:
We should ensure that `drain` should not create the objects and it specialized to do that.

Reply:
@harshit-gangal what do you think about this patch (passes tests, but I did not commit/push yet):
```diff
diff --git a/go/mysql/query.go b/go/mysql/query.go
index 18e9872752..003295b61e 100644
--- a/go/mysql/query.go
+++ b/go/mysql/query.go
@@ -39,6 +39,10 @@ var (
        ErrExecuteFetchMultipleResults = vterrors.Errorf(vtrpc.Code_INTERNAL, "unexpected multiple results. Use ExecuteFetchMulti instead.")
 )

+const (
+       FETCH_NO_ROWS = math.MinInt
+)
+
 //
 // Client side methods.
 //
@@ -322,7 +326,7 @@ func (c *Conn) ExecuteFetch(query string, maxrows int, wantfields bool) (result
 // caring for any results. The function returns an error if any of the statements fail.
 // The function drains the query results of all statements, even if there's an error.
 func (c *Conn) ExecuteFetchMultiDrain(query string) (err error) {
-       _, more, err := c.ExecuteFetchMulti(query, 0, false)
+       _, more, err := c.ExecuteFetchMulti(query, FETCH_NO_ROWS, false)
        return c.drainMoreResults(more, err)
 }

@@ -331,7 +335,7 @@ func (c *Conn) ExecuteFetchMultiDrain(query string) (err error) {
 func (c *Conn) drainMoreResults(more bool, err error) error {
        for more {
                var moreErr error
-               _, more, _, moreErr = c.ReadQueryResult(-1, false)
+               _, more, _, moreErr = c.ReadQueryResult(FETCH_NO_ROWS, false)
                err = errors.Join(err, moreErr)
        }
        return err
@@ -445,6 +449,10 @@ func (c *Conn) ReadQueryResult(maxrows int, wantfields bool) (*sqltypes.Result,
                }
        }

+       if maxrows == FETCH_NO_ROWS {
+               c.recycleReadPacket()
+               return result, more, warnings, nil
+       }
        // read each row until EOF or OK packet.
        for {
                data, err := c.readEphemeralPacket()
```

Reply:
Hmmm actually I still have errors.

Reply:
OK, this passes tests:
```diff
diff --git a/go/mysql/query.go b/go/mysql/query.go
index 18e9872752..d0f68ec35c 100644
--- a/go/mysql/query.go
+++ b/go/mysql/query.go
@@ -39,6 +39,13 @@ var (
        ErrExecuteFetchMultipleResults = vterrors.Errorf(vtrpc.Code_INTERNAL, "unexpected multiple results. Use ExecuteFetchMulti instead.")
 )

+const (
+       // Use as `maxrows` in `ExecuteFetch` and related functions, to indicate no rows should be fetched.
+       // This is different than specifying `0`, because `0` means "expect zero results", while this means
+       // "do not attempt to read any results into memory".
+       FETCH_NO_ROWS = math.MinInt
+)
+
 //
 // Client side methods.
 //
@@ -322,7 +329,7 @@ func (c *Conn) ExecuteFetch(query string, maxrows int, wantfields bool) (result
 // caring for any results. The function returns an error if any of the statements fail.
 // The function drains the query results of all statements, even if there's an error.
 func (c *Conn) ExecuteFetchMultiDrain(query string) (err error) {
-       _, more, err := c.ExecuteFetchMulti(query, 0, false)
+       _, more, err := c.ExecuteFetchMulti(query, FETCH_NO_ROWS, false)
        return c.drainMoreResults(more, err)
 }

@@ -331,7 +338,7 @@ func (c *Conn) ExecuteFetchMultiDrain(query string) (err error) {
 func (c *Conn) drainMoreResults(more bool, err error) error {
        for more {
                var moreErr error
-               _, more, _, moreErr = c.ReadQueryResult(-1, false)
+               _, more, _, moreErr = c.ReadQueryResult(FETCH_NO_ROWS, false)
                err = errors.Join(err, moreErr)
        }
        return err
@@ -451,6 +458,9 @@ func (c *Conn) ReadQueryResult(maxrows int, wantfields bool) (*sqltypes.Result,
                if err != nil {
                        return nil, false, 0, sqlerror.NewSQLError(sqlerror.CRServerLost, sqlerror.SSUnknownSQLState, "%v", err)
                }
+               if maxrows == FETCH_NO_ROWS {
+                       return result, more, warnings, nil
+               }

                if c.isEOFPacket(data) {
                        defer c.recycleReadPacket()
```

Reply:
Committed as https://github.com/vitessio/vitess/pull/14949/commits/a30c8045ff79dcd5fede6a27a57dc7ac48327411, let's see how the entire CI reacts.

Reply:
CI looks happy. @harshit-gangal what do you think?

Reply:
It looks good to me